can/io/mf4.py

```diff
@@ -0,0 +1,160 @@
+# coding: utf-8
+
+"""
+Contains handling of MF4 logging files.
+
+"""
+
+from __future__ import absolute_import
+
+from datetime import datetime
+from pathlib import Path
+import logging
+
+from ..listener import Listener
+from ..util import channel2int
+from .generic import BaseIOHandler
+
+from asammdf import MDF, Signal
+import numpy as np
+
+CAN_MSG_EXT = 0x80000000
+CAN_ID_MASK = 0x1FFFFFFF
+
+DTYPE = np.dtype(
+    [
+        ('BusChannel', '<u1'),
+        ('ID', '<u4'),
+        ('IDE', '<u1'),
+        ('DLC', '<u1'),
+        ('DataLength', '<u1'),
+        ('DataBytes', '(64,)u1'),
+        ('Dir', '<u1'),
+        ('EDL', '<u1'),
+        ('BRS', '<u1'),
+        ('ESI', '<u1'),
+    ]
+)
+
+FD_LEN2DLC = {
+    12: 9,
+    16: 10,
+    20: 11,
+    24: 12,
+    32: 13,
+    48: 14,
+    64: 15,
+}
+
+logger = logging.getLogger('can.io.mf4')
+
+
+class MF4Writer(BaseIOHandler, Listener):
+    """Logs CAN data to an ASAM Measurement Data File v4 (.mf4).
+
+    If a message has a timestamp smaller than the previous one or None,
+    it gets assigned the timestamp that was written for the last message.
+    It the first message does not have a timestamp, it is set to zero.
+    """
+
+    def __init__(self, file, database=''):
+        """
+        :param file: a path-like object or as file-like object to write to
+                     If this is a file-like object, is has to opened in binary
+                     write mode, not text write mode.
+        :param database: optional path to a DBC or ARXML file that contains
+                         message description.
+        """
+        super(MF4Writer, self).__init__(file, mode='r+b')
+        now = datetime.now()
+        self._mdf = MDF()
+        self._mdf.header.start_time = now
+        self.last_timestamp = self._start_time = now.timestamp()
+
+        if database:
+            database = Path(database).resolve()
+            if database.exists():
+                data = database.read_bytes()
+                attachment = data, database.name
+            else:
+                attachment = None
+        else:
+            attachment = None
+
+        sigs = [
+            Signal(
+                name="CAN_DataFrame",
+                samples=np.array([], dtype=DTYPE),
+                timestamps=np.array([], dtype='<f8'),
+                attachment=attachment,
+            )
+        ]
+        self._mdf.append(sigs)
+
+        sigs = [
```

Comment:
I think you accidentally duplicated something ;)

Reply:
This is necessary to correctly store normal and error frames

https://github.com/danielhrisca/python-can/blob/mf4_io/can/io/mf4.py#L86

https://github.com/danielhrisca/python-can/blob/mf4_io/can/io/mf4.py#L96

Reply:
Ah, now I see it. I think it would be clearer if you wrote it like that:

```python
self._mdf.append([
    Signal(
        name="CAN_DataFrame",
        ...
    ),
    Signal(
        name="CAN_ErrorFrame",
        ...
    )
])
```

Reply:
I've made another simplification. The signals have to be appended separately to create individual channel groups for each of them

Reply:
Looks good!